src/hooks/vitest.ts

```diff
@@ -0,0 +1,197 @@
+import assert from "node:assert";
+import { pathToFileURL } from "node:url";
+import worker from "node:worker_threads";
+
+import { simple as walk } from "acorn-walk";
+import { type ESTree } from "meriyah";
+
+import { args as args_, assignment, call_, identifier, literal, member, ret } from "../generate";
+import { info, warn } from "../message";
+import { recording, start } from "../recorder";
+import Recording from "../Recording";
+import genericTransform from "../transform";
+
+function createInitChannel() {
+  return new worker.BroadcastChannel("appmap-node/vitest/initialized");
+}
+
+// There can be 3 processes (p1, p2, p3) and many threads involved depending on the
+// node version and vitest threading arguments. Each thread will load this file so we have to
+// determine in which process and thread we are going to listen for initialization message.
+// Omitting this ugly check and not preventing "innocent" creation of the initChannel in wrong
+// processes and threads results in issues (process hangs). Also, we want to avoid listening in
+// more than one thread to prevent multiple info messages.
+// p1: node .../.yarn/releases/yarn-3.6.3.cjs vitest run [--no-threads] [--single-thread]
+// p2: node .../vitest/vitest.mjs run [--no-threads] [--single-thread]
+// p3: node .../vitest/dist/child.js (this process exists only in --no-threads mode)
+function shouldListenForInit() {
+  const isNoThreadsModeMainThread =
+    worker.isMainThread && process.argv.length > 1 && process.argv[1].endsWith("child.js");
+  const [major, ,] = process.versions.node.split(".").map(Number);
+  // There will be 2 threads (#0, #1) in [--no-threads, node 20], so we don't want
+  // to listen in thread #1 for other modes in this case. On the other hand,
+  // we don't have thread #2 in [--single-thread node 18/16], but we can listen
+  // in thread #1 because we don't have #1 in [--no-threads node 18/16].
+  // Thus we guarantee that a single thread listens in all node versions and thread modes combinations.
+  const threadIdToListenInOtherModes = major >= 20 ? 2 : 1;
+  return isNoThreadsModeMainThread || worker.threadId === threadIdToListenInOtherModes;
+}
```

Comment:
Ouch! I wish this hack wasn't necessary. Great job figuring it out and thank you for detailed comments! Fingers crossed it doesn't break anytime soon :)